inc/zoo/swar/SWAR.h

```diff
@@ -74,57 +78,152 @@ template<int NBits, typename T = uint64_t> struct SWAR {
+// of your MSBs are set 0 (leaving space for the operation).  Similarly, doing
+// multiplications via SWAR techniques require double bits per lane (unless you
+// can bind your inputs at half lane size).
+// This leads to a useful technique (which we use in the robin hood table)
+// where we interleave two related small bit count integers inside of a lane of
+// swar.  More generally, this is useful because it sometimes allows fast
+// operations on side "a" of some lane if side "b" is blitted out, and vice
+// versa.  In the spirit of separation of concerns, we provide a cut-lane-SWAR
+// abstraction here.
+
+template<int NBitsMost, int NBitsLeast, typename T = uint64_t> struct SWARWithSubLanes : SWAR<NBitsMost+NBitsLeast, T> {
+    static constexpr inline auto Available = sizeof(T);
+    static constexpr inline auto LaneBits = NBitsLeast+NBitsMost;
+    static constexpr inline auto NSlots = Available * 8 / LaneBits;
+
+    SWARWithSubLanes() = default;
+    constexpr explicit SWARWithSubLanes(T v) : SWAR<NBitsMost+NBitsLeast, T>(v) {}
+    constexpr explicit operator T() const noexcept { return this->m_v; }
+
+    // M is most significant bits slice, L is least significant bits slice.
+    // 0x....M2L2M1L1 or MN|LN||...||M2|L2||M1|L1
+    using SL = SWARWithSubLanes<NBitsMost, NBitsLeast, T>;
+    using Base = swar::SWAR<LaneBits, T>;
+
+    //constexpr T Ones = meta::BitmaskMaker<NBits, SWAR<NBits, T>{1}.value(), T>::value;
+    static constexpr inline auto LeastOnes = meta::BitmaskMaker<T, Base{1}.value(), LaneBits>::value;
+    static constexpr inline auto MostOnes = meta::BitmaskMaker<T, Base{1<<NBitsLeast}.value(), LaneBits>::value;
+    static constexpr inline auto LeastMask = meta::BitmaskMaker<T, Base{~(~0ull<<NBitsLeast)}.value(), LaneBits>::value;
+    static constexpr inline auto MostMask = ~LeastMask;
+
+    constexpr auto least() const noexcept {
+        return this->m_v & LeastMask;
+    }
+
+    // Returns only the least significant bits at specified position.
+    constexpr auto least(u32 pos) const noexcept {
+        constexpr auto filter = (T(1) << LaneBits) - 1;
+        const auto keep = (filter << (LaneBits * pos)) & LeastMask;
+        return this->m_v & keep;
+    }
+
+    // Returns only the least significant bits at specified position, 'decoded' to their integer value.
+    constexpr auto leastFlat(u32 pos) const noexcept {
+        return least(pos) >> (LaneBits*pos);
+    }
+
+    constexpr auto most() const noexcept {
+        return this->m_v & MostMask;
+    }
+
+    // Returns only the most significant bits at specified position.
+    constexpr auto most(u32 pos) const noexcept {
+        constexpr auto filter = (T(1) << LaneBits) - 1;
+        const auto keep = (filter << (LaneBits * pos)) & MostMask;
+        return this->m_v & keep;
+    }
+
+    // Returns only the most significant bits at specified position, 'decoded' to their integer value.
+    constexpr auto mostFlat(u32 pos) const noexcept {
```

Comment:
there is a missing operation here, `flattenMost`, which blits the most significant bits into the leasts

Reply:
Agree, useful op, but also needs HUGE WARNING about there being zero guarantee the sublanes will be the same size.